contracts/root/predicates/ERC20Predicate.sol

```diff
@@ -0,0 +1,299 @@
+pragma solidity ^0.5.2;
+
+import { BytesLib } from "../../common/lib/BytesLib.sol";
+import { Common } from "../../common/lib/Common.sol";
+import { Math } from "openzeppelin-solidity/contracts/math/Math.sol";
+import { RLPEncode } from "../../common/lib/RLPEncode.sol";
+import { RLPReader } from "solidity-rlp/contracts/RLPReader.sol";
+
+import { IErcPredicate } from "./IPredicate.sol";
+import { WithdrawManagerHeader } from "../withdrawManager/WithdrawManagerStorage.sol";
+
+contract ERC20Predicate is IErcPredicate {
+  using RLPReader for bytes;
+  using RLPReader for RLPReader.RLPItem;
+
+  bytes32 constant DEPOSIT_EVENT_SIG = 0x4e2ca0515ed1aef1395f66b5303bb5d6f1bf9d61a353fa53f73f8ac9973fa9f6;
+  bytes32 constant WITHDRAW_EVENT_SIG = 0xebff2602b3f468259e1e99f613fed6691f3a6526effe6ef3e768ba7ae7a36c4f;
+  bytes32 constant LOG_TRANSFER_EVENT_SIG = 0xe6497e3ee548a3372136af2fcb0696db31fc6cf20260707645068bd3fe97f3c4;
+  // 0x2e1a7d4d = keccak256('withdraw(uint256)').slice(0, 4)
+  bytes4 constant WITHDRAW_FUNC_SIG = 0x2e1a7d4d;
+  // 0xa9059cbb = keccak256('transfer(address,uint256)').slice(0, 4)
+  bytes4 constant TRANSFER_FUNC_SIG = 0xa9059cbb;
+
+  constructor(address _withdrawManager) public IErcPredicate(_withdrawManager) {}
+
+  function startExitWithBurntTokens(bytes calldata data)
+    external
+  {
+    RLPReader.RLPItem[] memory referenceTxData = data.toRlpItem().toList();
+    bytes memory receipt = referenceTxData[6].toBytes();
+    RLPReader.RLPItem[] memory inputItems = receipt.toRlpItem().toList();
+    uint256 age = withdrawManager.verifyInclusion(data, 0 /* offset */, false /* verifyTxInclusion */);
+    uint256 logIndex = referenceTxData[9].toUint();
+    require(logIndex < MAX_LOGS, "Supporting a max of 10 logs");
+    inputItems = inputItems[3].toList()[logIndex].toList(); // select log based on given logIndex
+
+    // "address" (contract address that emitted the log) field in the receipt
+    address childToken = RLPReader.toAddress(inputItems[0]);
+    bytes memory logData = inputItems[2].toBytes();
+    inputItems = inputItems[1].toList(); // topics
+    // now, inputItems[i] refers to i-th (0-based) topic in the topics array
+    // event Withdraw(address indexed token, address indexed from, uint256 amountOrTokenId, uint256 input1, uint256 output1)
+    require(
+      bytes32(inputItems[0].toUint()) == WITHDRAW_EVENT_SIG,
+      "Not a withdraw event signature"
+    );
+    address rootToken = address(RLPReader.toUint(inputItems[1]));
+    require(
```

Comment:
why aren't we recovering `msg.sender` from raw tx sig. @atvanguard

Reply:
also why `false` in `verifyInclusion`?

Reply:
> why aren't we recovering msg.sender from raw tx sig

As discussed, including the tx sig is inconsequential :)

> also why false in verifyInclusion?

`false` is for `verifyTxInclusion` - If this flag is set, then it ALSO checks the inclusion of tx, receipt inclusion is *checked in any case*. Since we don't need to send the raw tx, we are not checking it's inclusion.